src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -691,54 +701,86 @@
+   (replace-catalog! catalog (now)))
+  ([{:keys [producer_timestamp resources certname] :as catalog} :- catalog-schema
     received-timestamp :- pls/Timestamp]
+   (time! (:replace-catalog performance-metrics)
+          (jdbc/with-db-transaction []
+            (let [hash (time! (:catalog-hash performance-metrics)
+                              (shash/catalog-similarity-hash catalog))
+                  {catalog-id :catalog_id
+                   stored-hash :catalog_hash
+                   certname-id :certname_id
+                   latest-producer-timestamp :producer_timestamp} (latest-catalog-metadata certname)]
+              (cond
+                (some-> latest-producer-timestamp
+                        (.after (to-timestamp producer_timestamp)))
+                (log/warnf "Not replacing catalog for certname %s because local data is newer." certname)
+                (= stored-hash hash)
+                (update-existing-catalog catalog-id hash catalog received-timestamp)
+                :else
+                (let [refs-to-hashes (time! (:resource-hashes performance-metrics)
+                                            (kitchensink/mapvals shash/resource-identity-hash resources))]
+                  (if (nil? catalog-id)
+                    (add-new-catalog certname-id hash catalog refs-to-hashes received-timestamp)
+                    (replace-existing-catalog certname-id catalog-id hash catalog refs-to-hashes received-timestamp))))
+              hash)))))
+
+(pls/defn-validated store-catalog! [catalog :- catalog-schema
+                                    received-timestamp :- pls/Timestamp]
+  (if @store-catalogs-historically?
+    (add-catalog! catalog received-timestamp)
+    (replace-catalog! catalog received-timestamp)))
 
-   (let [refs-to-hashes (time! (:resource-hashes performance-metrics)
-                               (reduce-kv (fn [acc k v]
-                                            (assoc acc k (shash/resource-identity-hash v)))
-                                          {} resources))
-         hash (time! (:catalog-hash performance-metrics)
-                     (shash/catalog-similarity-hash catalog))
-         {id :id
-          stored-hash :hash} (catalog-metadata certname)]
-
-     (jdbc/with-db-transaction []
-      (cond
-       (nil? id)
-       (add-new-catalog hash catalog refs-to-hashes received-timestamp)
-
-       (= stored-hash hash)
-       (update-catalog-hash-match id hash catalog received-timestamp)
-
-       :else
-       (update-catalog-hash-miss id hash catalog refs-to-hashes received-timestamp)))
-
-     hash)))
 
 (defn delete-catalog!
-  "Remove the catalog identified by the following hash"
-  [catalog-hash]
-  (jdbc/delete! :catalogs [(str (sutils/sql-hash-as-str "hash") "=?")
-                           catalog-hash]))
+  "Remove the catalog for a given certname.
+  Note that this function is only used in tests."
```

Comment:
It's also misleadingly documented in repl.markdown. 

But seriously, it looks like this fn exists solely for the purpose of being tested.